From dba1d480a2e01696442a7fee482c69c1c99c2bae Mon Sep 17 00:00:00 2001
From: Johannes Berg <johannes.berg@intel.com>
Date: Fri, 18 Nov 2022 17:06:22 +0530
Subject: [PATCH] wifi: mac80211: properly implement MLO key handling

Implement key installation and lookup (on TX and RX)
for MLO, so we can use multiple GTKs/IGTKs/BIGTKs.

Co-authored-by: Ilan Peer <ilan.peer@intel.com>
Signed-off-by: Ilan Peer <ilan.peer@intel.com>
Signed-off-by: Johannes Berg <johannes.berg@intel.com>
---
 include/net/mac80211.h     |   2 +
 net/mac80211/cfg.c         |  75 ++++++++++++---
 net/mac80211/ieee80211_i.h |   1 +
 net/mac80211/iface.c       |   9 ++
 net/mac80211/key.c         | 190 +++++++++++++++++++++++++------------
 net/mac80211/key.h         |  13 ++-
 net/mac80211/rx.c          |  44 +++++----
 net/mac80211/tx.c          |  56 +++++++++--
 8 files changed, 286 insertions(+), 104 deletions(-)

--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -169,6 +169,7 @@ struct ieee80211_tx_data {
 	struct ieee80211_local *local;
 	struct ieee80211_sub_if_data *sdata;
 	struct sta_info *sta;
+	struct link_sta_info *link_sta;
 	struct ieee80211_key *key;
 	struct ieee80211_tx_rate rate;
 
